keps/sig-auth/2718-20210511-client-exec-proxy/README.md

```diff
@@ -0,0 +1,845 @@
+      of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+* **Does enabling the feature change any default behavior?**
+  Any change of default behavior may be surprising to users or break existing
+  automations, so be extremely careful here.
+
+  No.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back
+  the enablement)?**
+  Also set `disable-supported` to `true` or `false` in `kep.yaml`.  Describe
+  the consequences on existing workloads (e.g., if this is a runtime feature,
+  can it break the existing applications?).
+
+  Yes, the feature can be disabled.  In the case where it is being used by long
+  running clients to talk to the API server, the client would have to be
+  reconfigured to stop using the proxy for connections.  Depending on how the
+  client is configured, this may require a restart.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+
+  The feature can be reenabled after being disabled without consequence.
+
+* **Are there any tests for feature enablement/disablement?**
+  The e2e framework does not currently support enabling or disabling feature
+  gates. However, unit tests in each component dealing with managing data, created
+  with and without the feature, are necessary. At the very least, think about
+  conversion tests if API types are being modified.
+
+  No.
+
+### Rollout, Upgrade and Rollback Planning
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How can a rollout fail? Can it impact already running workloads?**
+  Try to be as paranoid as possible - e.g., what if some components will restart
+   mid-rollout?
+
+* **What specific metrics should inform a rollback?**
+
+* **Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?**
+  Describe manual testing that was done and the outcomes.
+  Longer term, we may want to require automated upgrade/rollback tests, but we
+  are missing a bunch of machinery and tooling and can't do that now.
+
+* **Is the rollout accompanied by any deprecations and/or removals of features, APIs,
+fields of API types, flags, etc.?**
+  Even if applying deprecation policies, they may still surprise some users.
+
+### Monitoring Requirements
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How can an operator determine if the feature is in use by workloads?**
+  Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
+  checking if there are objects with field X set) may be a last resort. Avoid
+  logs or events for this purpose.
+
+* **What are the SLIs (Service Level Indicators) an operator can use to determine
+the health of the service?**
+  - [ ] Metrics
+    - Metric name:
+    - [Optional] Aggregation method:
+    - Components exposing the metric:
+  - [ ] Other (treat as last resort)
+    - Details:
```

Comment:
Number of requests?
Proxy latency?
Failed requests?
Lots of proxy restarts?

This code would be available to webhooks that interface with the API server so this remains relevant.

Reply:
Lots of proxy restarts seems reasonable. I'm not convinced there's much value in the others.

Reply:
As a client-side proxy component, how would you monitor this? Metrics for that specific proxy process would seem to be out of scope of this KEP.

Reply:
Maybe we can start with 1) number of failed requests and 2) number of proxy restarts.  @johnbelamaric there are other metrics in client-go so I think we can follow the same pattern. [1]  They would be exposed via a metrics endpoint in kube-controller-manager, for example, but not in kubectl.  

[1]: https://github.com/kubernetes/client-go/blob/f6ce18ae578c8cca64d14ab9687824d9e1305a67/tools/leaderelection/metrics.go

Reply:
+1 great.

Reply:
sorry, actually what I meant was "number of proxy restarts" - we are not deploying the actual proxy in this feature, right? We are just a client to it, via the domain socket? So how would we know if it restarted?

Reply:
@johnbelamaric in the existing credential plugin architecture, the binary actually invokes the plugin process: https://kubernetes.io/docs/reference/access-authn-authz/authentication/#client-go-credential-plugins

Same thing here.

Reply:
I guess it should be framed as "number of times we have to exec the proxy because of a connection failure"

Reply:
Looks like we already have a metric representing exec plugin calls: https://github.com/kubernetes/client-go/blob/d3b97581d2e22eef0f02a6667e19dfc193df7a3e/plugin/pkg/client/auth/exec/metrics.go#L93-L111